### ISSUE TITLE
Bug - MeshAgent fails to open url on ZorinOS 15.3 but reports "Successful"

### ISSUE THREAD

**Issue by LPJon** [Ylianst/MeshCentral#2881]
When using ZorinOS 15.3 Ultimate (Ubuntu Based) sending a command to the agent to open a url, either manually or through an MC button, does in fact not open a url even though the agent reports "Successful". I have already set the default web browser to FireFox in the agent OS. This means that the "Chat" button also does not open a "Chat" browser window on the agent side.

This a Location Deployment
MC v0.8.75
Agent Version:
```
{
  "openssl": "1.1.1k",
  "duktape": "v2.6.0",
  "commitDate": "2021-03-25T23:58:15.000Z",
  "commitHash": "6c36f2fa6b5920ea73eb9ef478647b2277dec888",
  "compileTime": "17:25:02, Mar 25 2021"
}
```
Is this actually a bug or does this feature not work on some Linux distro's?

![openurl-success](https://user-images.githubusercontent.com/46533576/125139083-7f1a0680-e0d5-11eb-9c74-539c9a9ed08e.jpg)

**Comment by Ylianst** [comment]
This is one for Bryan... he works on the agent.

**Comment by LPJon** [comment]
I'm aware....Just didn't know if I should mention him or let you do that..lol

**Comment by krayon007** [comment]
I'll take a look. I have a zorin VM, but it's an older one. Linux should be much simpler for this function than windows. It sounds like the agent isn't able to determine the uid of the logged in user. I've seen issues like this on PuppyLinux, where the distro removed utilities like who and ID and made it so uid 0 is used everywhere.

Is your zorin 15.3 a vanilla install? (So I can create a similar configured VM)

**Comment by LPJon** [comment]
Yes this is a vanilla install of Zorin 15.3 Ultimate totally stock with no special customization. and MC is using a Domain Location style deployment. I can provide you with the ISO if needed....they charge for it.

**Comment by LPJon** [comment]
> 
> 
> I'll take a look. I have a zorin VM, but it's an older one. Linux should be much simpler for this function than windows. It sounds like the agent isn't able to determine the uid of the logged in user. I've seen issues like this on PuppyLinux, where the distro removed utilities like who and ID and made it so uid 0 is used everywhere.
> 
> Is your zorin 15.3 a vanilla install? (So I can create a similar configured VM)

I do see both "who" and "id" installed on this OS....just fyi. Something else that just crossed my mind. I had to install this agent as the user rather than root using the sudo command. Could this be why it's not working?

**Comment by LPJon** [comment]
> 
> 
> > I'll take a look. I have a zorin VM, but it's an older one. Linux should be much simpler for this function than windows. It sounds like the agent isn't able to determine the uid of the logged in user. I've seen issues like this on PuppyLinux, where the distro removed utilities like who and ID and made it so uid 0 is used everywhere.
> > Is your zorin 15.3 a vanilla install? (So I can create a similar configured VM)
> 
> I do see both "who" and "id" installed on this OS....just fyi. Something else that just crossed my mind. I had to install this agent as the user rather than root using the sudo command. Could this be why it's not working?

Just tried switching to an install done by root. Same issue persists.

**Comment by LPJon** [comment]
@krayon007 I found that this command `xdg-open http://google.com` manually opens google.com from the command prompt using terminal. Is is okay if I ask how meshagent does this and maybe what the url is or is that a really long answer?

**Comment by LPJon** [comment]
@krayon007 I might need to create another issue for this but I also just noticed that the Clipboard "Get Clipboard" or "Set Clipboard" functions also do not work. This does seem to point more at what you were saying. Something about the base command looking for a user id or who to run the command as.

**Comment by LPJon** [comment]
@krayon007 The "Message" button DOES work! This distro seems to be based on the Ubuntu "Bionic" distro.

![Zorin OS Version](https://user-images.githubusercontent.com/46533576/125146072-bdb9bc00-e0e9-11eb-9e38-6eb1c0970c83.jpg)

![Zorin OS Button Test](https://user-images.githubusercontent.com/46533576/125146195-36b91380-e0ea-11eb-8f54-f298083b1145.jpg)

**Comment by krayon007** [comment]
I assume you have xclip installed? I've never really tried installing the agent as a user on Linux. I'll take a look at what the ramifications would be in regards to this feature.

If you could email me a link where I can find the iso for 15.3 I can try to set that up as well, in case it's something specific to this distro/version.

In the meantime I'll play around with some of my bunches to see if I can force this failure to happen on Linux

**Comment by LPJon** [comment]
@krayon007 So "xclip" was NOT installed by default. I've installed that and testing again now.

In the mean time I've sent you a link to download ZorinOS 15.3 Ultimate Directly from their website by email with the subject "Issue #2881 on github MeshAgent".

**Comment by LPJon** [comment]
@krayon007 After installing "xclip" the clipboard functions are now working too! This is a step forward! Okay, so after installing "xclip" and trying to launch "Chat" again, I now get a popup message that says "Your firefox profile cannot be loaded. It may be missing or inaccessible" as pictured below.
![firefox Profile error](https://user-images.githubusercontent.com/46533576/125146981-68cc7480-e0ee-11eb-9ca4-f7715b16eb53.jpg)
This message did not popup before unless meshagent was installed as root. Now it's it IS poping up when installed as another user.

**Comment by LPJon** [comment]
@krayon007 Be sure to let me know you got the email.

**Comment by krayon007** [comment]
> @krayon007 Be sure to let me know you got the email.

It's downloading now. Thanks!

**Comment by LPJon** [comment]
@krayon007 Any news on this issue?

**Comment by krayon007** [comment]
Can you run this command from the console tab?
`eval sendConsoleText(require('user-sessions').consoleUid())`

I checked the code, all openurl does is use xdg-open, using the UID from the above command.

**Comment by LPJon** [comment]
@krayon007 It came back with "122". I will investigate a little more tomorrow.

**Comment by krayon007** [comment]
122 is not correct. Normal user UIDs are 1000 and greater. 122 looks like the Display Manager UID. You can verify that with the following console command:
`eval sendConsoleText(require('user-sessions').gdmUid)`

Normally, consoleUid will return the GDM UID, if it could not detect a logged in user.
I had to re-download the iso, becuase my previous download was interrupted... Tomorrow I'll look into setting up a VM, to see if I can reproduce this...

I'll look over the code that does the logged in user detection, and see if I can have you run a few other tests in the meantime..

**Comment by LPJon** [comment]
@krayon007 Ok...that is a better explanation of the commands. I did in fact not have a user logged in when I ran that command. Let me try both with a logged in user and I will report back.

**Comment by LPJon** [comment]
After Login with a user I get the following results:

`eval sendConsoleText(require('user-sessions').consoleUid())` returns `1000`
I should note that `1000` is the `root` user and NOT the user I am logged in as. That user has a profile folder called `.profile` in the home directory for loading by firefox that is required for firefox to work, hence the error message in the previous post picture above. Firefox is not vanilla for this install but is available to install through their app manager. 

`eval sendConsoleText(require('user-sessions').gdmUid)` returns `122`

Yes I can run any tests you would like. Just let me know what you would like.

**Comment by krayon007** [comment]
How did you setup your system to have that user setup? I just installed a vanilla zorin 15.3 setup, and it seems to behave like regular Ubuntu.

**Comment by LPJon** [comment]
@krayon007 I setup that user when I first installed Zorin as part of the setup process. I did switch the browser to firefox instead of the default browser. There was not special setup for that user. Does it still work when you install Firefox from the app manager.
